packages/mobile/src/web3/contracts.ts

```diff
@@ -1,128 +1,147 @@
-  }
-  return gethWallet
-}
+      if (contractKit || ipcProvider || gethWallet) {
+        throw new Error('Kit not properly destroyed')
+      }
 
-export function openContractKit(fornoMode: boolean) {
-  if (!contractKit) {
-    Logger.info(
-      `${tag}@openContractKit`,
-      `Initializing contractkit, platform: ${Platform.OS}, forno mode: ${fornoMode}`
-    )
-    const web3 = new Web3(fornoMode ? getHttpProvider(DEFAULT_FORNO_URL) : openIpcProvider())
-    const wallet = openWallet()
-    contractKit = newKitFromWeb3(web3, wallet)
+      const fornoMode = yield select(fornoSelector)
+      Logger.info(`${TAG}@initContractKit`, `Initializing contractkit, forno mode: ${fornoMode}`)
+
+      ipcProvider = getIpcProvider()
+      const web3 = new Web3(fornoMode ? getHttpProvider(DEFAULT_FORNO_URL) : ipcProvider)
+
+      Logger.info(`${TAG}@initContractKit`, 'Initializing wallet')
+      gethWallet = new RpcWallet(ipcProvider)
+      yield call([gethWallet, gethWallet.init])
+      Logger.info(
+        `${TAG}@initContractKit`,
+        `Initialized wallet with accounts: ${gethWallet.getAccounts()}`
+      )
+      contractKit = newKitFromWeb3(web3, gethWallet)
+      Logger.info(`${TAG}@initContractKit`, 'Initialized kit')
+      return
+    } catch (error) {
+      if (isProviderConnectionError(error)) {
+        retries -= 1
+        Logger.error(
+          `${TAG}@initContractKit`,
+          `Error initializing kit, could not connect to IPC. Retries remaining: ${retries}`,
+          error
+        )
+        if (retries <= 0) {
+          break
+        }
+
+        destroyContractKit()
+        yield delay(KIT_INIT_RETRY_DELAY)
+      } else {
+        Logger.error(`${TAG}@initContractKit`, 'Unexpected error initializing kit', error)
+        break
+      }
+    }
   }
-  return contractKit
+
+  Logger.error(`${TAG}@initContractKit`, 'Kit init unsuccessful, navigating to error screen')
+  navigateToError(ErrorMessages.CONTRACT_KIT_INIT_FAILED)
 }
 
-export function closeContractKit() {
-  Logger.debug(`closeContractKit`)
+export function destroyContractKit() {
+  Logger.debug(`${TAG}@closeContractKit`)
   contractKit = undefined
   gethWallet = undefined
   ipcProvider = undefined
 }
 
-export const getConnectedWalletAsync = async () => promisifyGenerator(getConnectedWallet())
-
-export function* getConnectedWallet() {
-  yield call(waitForGethConnectivity)
-  const wallet = openWallet()
-  if (!wallet.isSetupFinished()) {
-    yield call([wallet, wallet.init])
-    Logger.debug(`getConnectedWallet`, `Initialized wallet with accounts: ${wallet.getAccounts()}`)
+export function* getContractKit() {
+  if (!contractKit) {
+    yield call(initContractKit)
   }
-  return wallet
+  return contractKit
 }
 
-export const getContractKitAsync = async () => promisifyGenerator(getContractKit())
+// Used for cases where CK must be access outside of a saga
+export async function getContractKitAsync() {
+  let retries = 10
```

Comment:
I think it would be nice to abstract the retry logic so that we can swap it out with something nicer in future

Reply:
Yeah, there may be opportunities to DRY things up but it tends to be we have quite unique error handling + flow management in the major retry blocks (like web3 syncing)